src/operator/random/shuffle_op.cc

```diff
@@ -0,0 +1,132 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2018 by Contributors
+ * \file shuffle_op.cc
+ * \brief Operator to shuffle elements of an NDArray
+ */
+#if (__GNUC__ > 4 && !defined(__clang__major__)) || (__clang_major__ > 4 && __linux__)
+  #define USE_GNU_PARALLEL_SHUFFLE
+#endif
+
+#include <mxnet/operator_util.h>
+#include <algorithm>
+#include <random>
+#include <vector>
+#ifdef USE_GNU_PARALLEL_SHUFFLE
+  #include <parallel/algorithm>
+#endif
+#include "../elemwise_op_common.h"
+
+namespace mxnet {
+namespace op {
+
+namespace {
+
+  template<typename DType, typename Rand>
+  void Shuffle1D(DType* out, index_t size, Rand* prnd) {
+    #ifdef USE_GNU_PARALLEL_SHUFFLE
+      auto rand_n = [prnd](index_t n) {
+        std::uniform_int_distribution<index_t> dist(0, n - 1);
+        return dist(*prnd);
+      };
+      __gnu_parallel::random_shuffle(out, out + size, rand_n);
+    #else
+      std::shuffle(out, out + size, *prnd);
+    #endif
+  }
+
+  template<typename DType, typename Rand>
+  void ShuffleND(DType* out, index_t size, index_t first_axis_len, Rand* prnd) {
+    // Fisher-Yates shuffling
+    const index_t stride = size / first_axis_len;
+    auto rand_n = [prnd](index_t n) {
+      std::uniform_int_distribution<index_t> dist(0, n - 1);
+      return dist(*prnd);
+    };
+    for (index_t i = first_axis_len - 1; i > 0; --i) {
+      index_t j = rand_n(i + 1);
+      if (i != j) {
+        std::swap_ranges(out + stride * i, out + stride * (i + 1), out + stride * j);
+      }
+    }
+  }
+
+}  // namespace
+
+void ShuffleForwardCPU(const nnvm::NodeAttrs& attrs,
+                       const OpContext& ctx,
+                       const std::vector<TBlob>& inputs,
+                       const std::vector<OpReqType>& req,
+                       const std::vector<TBlob>& outputs) {
+  using namespace mxnet_op;
+  if (req[0] == kNullOp) {
+    return;
+  }
+  CHECK_NE(req[0], kAddTo) << "Shuffle does not support AddTo";
+  const TShape input_shape = inputs[0].shape_;
+  const index_t size = inputs[0].Size();
+  const index_t first_axis_len = input_shape[0];
+  Stream<cpu> *s = ctx.get_stream<cpu>();
+  MSHADOW_REAL_TYPE_SWITCH(inputs[0].type_flag_, DType, {
```

Comment:
What's the reason of not supporting integer types?

Reply:
Fixed with MSHADOW_TYPE_SWITCH